emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
```

Comment:
As with the `hashCode`, we could also provide a non-abstract, yet scalable implementation of this here: `(this minus that).isEmpty && (that minus this).isEmpty`.